keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2038 @@
+  included in the list that will be passed to the post-filter and to get copied
+  into the PotentialNodes field there.
+- A node can already be chosen while there is no information yet and, if
+  allocation for that node actually works, the Pod can get scheduled sooner.
+- Some resource drivers might not have any unsuitable nodes, for example
+  because they modify containers and that works on all nodes at all
+  times. Forcing such drivers to set an allow list would cause unnecessary
+  work.
+
+In its state for the Pod the scheduler plugin must remember when it rejected a
+node because of UnsuitableNodes. That information will be used in Post-filter
+to deallocate resources.
+
+#### Post-filter
+
+This is called when no suitable node could be found. If the Pod depends on ResourceClaims with delayed
+allocation, then deallocating one or more of these ResourceClaims may make the
+Pod schedulable after allocating the resource elsewhere. Therefore each
+ResourceClaim with delayed allocation is checked whether all of the following
+conditions apply:
+- allocated
+- not currently in use
+- it was the reason why some node could not fit the Pod, as recorded earlier in
+  Filter
+
+One of the ResourceClaims satisfying these criteria is picked randomly and freeing
+it is requested by setting the Deallocate field. This may make it possible to run the Pod
+elsewhere. If it still doesn't help, deallocation may continue with another
+ResourceClaim, if there is one. To prevent deallocating all resources merely
+because the scheduler happens to check quickly, the next deallocation will only
+requested when there is none currently running.
+
+At the moment, the scheduler has no information that might enable it to
+prioritize which resource to deallocate first. Future extensions of this KEP
+might attempt to improve this.
+
+#### Pre-score
+
+This is passed a list of nodes that have passed filtering by the resource
+plugin and the other plugins. The PotentialNodes field of unallocated
+ResourceClaims with delayed allocation gets updated now if the field doesn't
+match the current list already and there is no scheduler extender for the
+claim.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If using delayed allocation and the resource has not been allocated yet,
+the SelectedNode field of the ResourceClaim
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+If all resources have been allocated already,
+the scheduler adds the Pod to the ReservedFor field of its ResourceClaims to ensure that
+no-one else gets to use those.
+
+If some resources are not allocated yet or reserving an allocated resource
+fails, the scheduling attempt needs to be aborted and retried at a later time
+or when the ResourceClaims change.
+
+#### Unreserve
+
+The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
+all.
+
+### Cluster Autoscaler
```

Comment:
@MaciekPytel @x13n does this make sense?

Reply:
@MaciekPytel's comment in https://github.com/kubernetes/enhancements/pull/3371#discussion_r893584762 is applicable here:
> I'm not saying we need a detailed design here, but I think we should at least have a general idea that we all relevant sigs agree is feasible.

So we just need to check whether this idea here seems feasible.

Reply:
So the proposal seems to be for drivers to expose a new API just for CA? That means each driver has to be autoscaling-ready instead of just being k8s-ready? How can we prevent the behavior of separate API surfaces for CA and scheduler from diverging?

Reply:
> So the proposal seems to be for drivers to expose a new API just for CA?

Yes, because allocating a resource and simulating the allocation are different things. The autoscaler cannot use the API for allocating a resource because that would interfere with actually using the resource while the autoscaler is doing a simulation run.

> That means each driver has to be autoscaling-ready instead of just being k8s-ready?

Yes. I understand that this isn't ideal, but I just don't see a way around that :cry: 

Limiting the capabilities of the scheduler to what can be supported without extra work in or for autoscaler seems backwards to me: not all users of the scheduler are also users of the autoscaler.

@kate-goldenring: in Akri, do you you support cluster autoscaling? If not, has such support been requested and/or considered?

> How can we prevent the behavior of separate API surfaces for CA and scheduler from diverging?

I expect that we'll continue to evolve both at the same time. I already have generic utility code for the scheduler interaction in my prototype. Support for the CA API can be added there, together with tests that resource driver developers can use to verify their implementation (similar to csi-sanity testing).

Reply:
@pohly I have not tried Akri with CA nor heard of anyone considering it; though, I'd expect it would work. Since Akri uses device plugins, our agent assumes the kubelet is the source of truth, so if kubelet tries to run a Pod that uses a resource, Akri will allocate it (unless the resources capacity has been used up). I will try to investigate this further.

Reply:
CA has special support for extended resources backed by GPUs (https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/processors/customresources/gpu_processor.go). I don't know whether the same would be needed for Akri.

Reply:
The main thing that this code does is to consider nodes as not ready until GPU driver adds a GPU resource. We can certainly have a more generic solution to support other ways of initializing nodes. Scheduling simulation itself is handled by scheduler framework, since GPUs are then just a part of node allocatable.

I guess if there's no way for scheduler to decide on its own whether a pod can get scheduled on a given node without consulting the driver, then CA has to consult them as well. However, it would be useful to converge API surface as much as possible to avoid hard to debug autoscaling issues.

An alternative would be to push more information from drivers to scheduler via k8s objects, instead of just having one number which is node allocatable. That way drivers would be responsible for publishing this information and scheduler & CA would just operate on standardized k8s objects instead of relying on an API call. Why is this not feasible, are the different resources just too diverse to be represented with a manageable set of models?

Reply:
The other complication with Akri and CA might be that Akri manages network-attached devices that get reported as local resources on nodes and uses a scheduler extender. CA does not support scheduler extenders. When relying exclusively on resource counts, CA might incorrectly conclude that adding a new node will allow a pod to run that is currently waiting for such a device although adding nodes won't really make a new device available. Then the pod is still stuck and the extra node will just get scaled down again. It all depends on what a fictional node will look like. The other error case is when it doesn't scale up because network-attached devices are incorrectly assumed to be not available on fictional nodes. I'm just trying to point out that it is not guaranteed to work correctly just because Akri uses the device plugin API.

The more important aspect is that for users of Akri, the dynamic resource API seems useful even if some future resource drivers for it don't support autoscaling.

> Why is this not feasible, are the different resources just too diverse to be represented with a manageable set of models?

Yes. This is [has been tried before](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc/edit#) without reaching a consensus what such models should look like.

Reply:
> uses a scheduler extender

For scheduler extender, are you referring to Akri's controller? The use of the controller is optional. It just eases deployments if desired.

> CA might incorrectly conclude that adding a new node will allow a pod to run that is currently waiting for such a device although adding nodes won't really make a new device available

I am still ramping up on CA, but is this a built-in assumption of CA that adding nodes adds more of the resource? It seems like that GPU extension for CA could be made generic for all extended resources, checking which nodes have any allocatable resources before making decisions.

Reply:
> For scheduler extender, are you referring to Akri's controller? The use of the controller is optional. It just eases deployments if desired.

Yes, I mixed that up. I thought a scheduler extender was involved somehow to avoid scheduling too many pods when they compete for the same device. Let me give an example. Suppose there are network-attached devices X and Y of type foo and nodes M, N and J. Device X is available on nodes M and N. Y is only available on J. Akri will then report having one device of type foo on each node, right?

Now suppose two pods A and B get created which both need one device of type foo. Can it happen that pod A gets scheduled onto node M and pod B onto node N? To the scheduler it must look like both these nodes have enough resources, so I suppose it can happen. The pod that starts first grabs device X and the other one is stuck. The right scheduling decision would have been to put one pod onto node J where it then can use device Y.

I thought Akri somehow prevented such bad scheduling decisions and couldn't quite understand how that can work without race conditions when using the device plugin API, but perhaps it simply doesn't prevent them in all cases?

If I got this right, then this KEP would improve this situation by never scheduling a pod onto a node unless the resources that it needs really are available and reserved for it there.

> is this a built-in assumption of CA that adding nodes adds more of the resource

Resources reported through kubelet are assumed to be local to the node, so yes, CA assumes that adding more nodes will also add more extended resources.

Reply:
> Suppose there are network-attached devices X and Y of type foo and nodes M, N and J. Device X is available on nodes M and N. Y is only available on J. Akri will then report having one device of type foo on each node, right?

What does "available" mean? In my explanation, I'm assuming it means that M and N can see/discover device X while only J can see/discover Y, but happy to re-respond if its different. In this case, Akri creates device plugins on nodes M and N for device X and a device plugin on Y for J. I am not sure which of the lettered devices are "foo" but say with Akri, you configured it so that at most one container can use a device at once -- the discovered devices have a `capacity` of 1 in the Akri Configuration. The device plugins on both nodes M and N show that device X (akri.sh/device-x-id) has one available allocation. You deploy Pod A and Pod B (say with node taints to be scheduled to nodes M and N) to use device X. the kubelet on Node M calls allocate on the Akri agent on its node. The Agent looks up the Akri Instance custom resource for device X and sees its one slot is available. It reserves it. Simultaneously, the kubelet on node N tries to update the instance. The API server will return an error due to the Instance update being out of sync (probably something along the lines of the `generation` of the Instance being out of sync). On the error, the Agent pulls the latest instance, notices the slot was reserved, fails the allocation request, and updates kubelet that the device is not available (`[Unhealthy]`). Pod will go into a Pending state saying that the required resources are not available. This is expected. A user could increase the `capacity` in the Akri Configuration to enable more nodes/pods to use the device.

More information about this can be found in our architecture docs: https://docs.akri.sh/architecture/architecture-overview

Reply:
I do see the point that ideally only one Pod would ever attempt to be run by kubelet. There is a temporary error state that is resolved.

Reply:
@MaciekPytel @x13n what are your thoughts on the evolution of the KEP? This still requires your approval.